src/Stratis.Features.Collateral/CollateralFederationManager.cs

```diff
@@ -211,5 +220,64 @@ public CollateralFederationMember CollateralAddressOwner(VotingManager votingMan
 
             return member;
         }
+
+        private T BinaryFindFirst<T>(T[] array, Func<T, bool?> func, int first = 0, int? span = null)
+        {
+            int length = span ?? array.Length;
+
+            // If the last item does not fit the criteria then don't bother looking any further.
+            if (length == 0 || func(array[first + length - 1]) == false)
+                return default;
+
+            // If the first item matches the criteria then take it and don't bother looking beyond it.
+            if (func(array[first]) == true)
+                return array[first];
+
+            // If there are no other items then return.
+            if (length <= 2)
+                return default;
+
+            int pivot = length / 2;
+
+            return BinaryFindFirst(array, func, first + 1, pivot - 1)
+                ?? BinaryFindFirst(array, func, first + pivot, length - pivot - 1);
+        }
+
+        /// <inheritdoc />
+        public override int? GetMultisigMinersApplicabilityHeight()
+        {
+            IConsensusManager consensusManager = this.fullNode.NodeService<IConsensusManager>();
+            ChainedHeader fork = (this.lastBlockChecked == null) ? null : consensusManager.Tip.FindFork(this.lastBlockChecked);
+
+            if (this.multisigMinersApplicabilityHeight != null && fork?.HashBlock == this.lastBlockChecked?.HashBlock)
+                return this.multisigMinersApplicabilityHeight;
+
+            this.lastBlockChecked = fork;
+            this.multisigMinersApplicabilityHeight = null;
+            var commitmentHeightEncoder = new CollateralHeightCommitmentEncoder(this.logger);
+
+            ChainedHeader[] headers = consensusManager.Tip.EnumerateToGenesis().TakeWhile(h => h != this.lastBlockChecked).Reverse().ToArray();
+
+            ChainedHeader first = BinaryFindFirst<ChainedHeader>(headers, (chainedHeader) =>
```

Comment:
What is the purpose of using `BinaryFindFirst`? Is it for performance? If so, please move to Stratis.Bitcoin somewhere as its more of a"Core" method.

Reply:
Moved.